src/pipelinedefinitionstatus.hpp

```diff
@@ -0,0 +1,288 @@
+        return {};
+    }
+    StateKeeper handle(const ValidationFailedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<LoadingFailedLastValidationRequiredRevalidation> handle(const UsedModelChangedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<RetiredState> handle(const RetireEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+};
+
+struct LoadingFailedLastValidationRequiredRevalidation {
+    static const PipelineDefinitionStateCode code = PipelineDefinitionStateCode::LOADING_PRECONDITION_FAILED_REQUIRED_REVALIDATION;
+    PipelineDefinitionStateCode getState() const {
+        return code;
+    }
+    void print() const {
+        SPDLOG_ERROR(pipelineDefinitionStateCodeToString(getState()));
+    }
+    StateChanger<AvailableState> handle(const ValidationPassedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<LoadingPreconditionFailedState> handle(const ValidationFailedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateKeeper handle(const UsedModelChangedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<RetiredState> handle(const RetireEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+};
+
+struct RetiredState {
+    static const PipelineDefinitionStateCode code = PipelineDefinitionStateCode::RETIRED;
+    PipelineDefinitionStateCode getState() const {
+        return code;
+    }
+    void print() const {
+        SPDLOG_ERROR(pipelineDefinitionStateCodeToString(getState()));
+    }
+    StateChanger<AvailableState> handle(const ValidationPassedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<LoadingPreconditionFailedState> handle(const ValidationFailedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateKeeper handle(const UsedModelChangedEvent& e) const {
+        SPDLOG_ERROR("State:{} handling: {}. {}", pipelineDefinitionStateCodeToString(getState()), e.name, INVALID_TRANSITION_MESSAGE);
+        throw std::logic_error(INVALID_TRANSITION_MESSAGE);
+        return {};
+    }
+    StateKeeper handle(const RetireEvent& e) const {
```

Comment:
Could we have base class implemented `handle` with std::logic_error so if we do not override it then we implicitly ensure the transition is unavailable?

Reply:
We want to avoid inheritance,